src/envs/satellites.py

```diff
@@ -0,0 +1,598 @@
+        self._HasCamera = Predicate("HasCamera", [self._sat_type],
+                                    self._HasCamera_holds)
+        self._HasInfrared = Predicate("HasInfrared", [self._sat_type],
+                                      self._HasInfrared_holds)
+        self._HasGeiger = Predicate("HasGeiger", [self._sat_type],
+                                    self._HasGeiger_holds)
+        self._ShootsChemX = Predicate("ShootsChemX", [self._sat_type],
+                                      self._ShootsChemX_holds)
+        self._ShootsChemY = Predicate("ShootsChemY", [self._sat_type],
+                                      self._ShootsChemY_holds)
+        self._HasChemX = Predicate("HasChemX", [self._obj_type],
+                                   self._HasChemX_holds)
+        self._HasChemY = Predicate("HasChemY", [self._obj_type],
+                                   self._HasChemY_holds)
+        self._CameraReadingTaken = Predicate("CameraReadingTaken",
+                                             [self._sat_type, self._obj_type],
+                                             self._CameraReadingTaken_holds)
+        self._InfraredReadingTaken = Predicate(
+            "InfraredReadingTaken", [self._sat_type, self._obj_type],
+            self._InfraredReadingTaken_holds)
+        self._GeigerReadingTaken = Predicate("GeigerReadingTaken",
+                                             [self._sat_type, self._obj_type],
+                                             self._GeigerReadingTaken_holds)
+
+        # Options
+        self._MoveTo = utils.SingletonParameterizedOption(
+            "MoveTo",
+            types=[self._sat_type, self._obj_type],
+            params_space=Box(0, 1, (2, )),  # target absolute x/y
+            policy=self._MoveTo_policy)
+        self._Calibrate = utils.SingletonParameterizedOption(
+            "Calibrate",
+            types=[self._sat_type, self._obj_type],
+            policy=self._Calibrate_policy)
+        self._ShootChemX = utils.SingletonParameterizedOption(
+            "ShootChemX",
+            types=[self._sat_type, self._obj_type],
+            policy=self._ShootChemX_policy)
+        self._ShootChemY = utils.SingletonParameterizedOption(
+            "ShootChemY",
+            types=[self._sat_type, self._obj_type],
+            policy=self._ShootChemY_policy)
+        self._UseInstrument = utils.SingletonParameterizedOption(
+            "UseInstrument",
+            types=[self._sat_type, self._obj_type],
+            policy=self._UseInstrument_policy)
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "satellites"
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        # Note: target_sat_x and target_sat_y are only used if we're not
+        # doing calibration, shooting Chemical X or Y, or using an instrument.
+        cur_sat_x, cur_sat_y, obj_x, obj_y, target_sat_x, target_sat_y, \
```

Comment:
what's the reason to use xy to identify entities if those entities have IDs?

Reply:
only the objects (the things that aren't satellites, that is) have IDs. also, i was lazy with the action space, i didn't want to make it 0 to inf for that dimension but 0 to 1 for everything else.